http-gateways/IPLD_GATEWAY.md

```diff
@@ -0,0 +1,196 @@
+# IPLD Gateway Specification
+
+**Authors**:
+
+- Mauve Signweaver ([@RangerMauve](https://github.com/RangerMauve))
+
+----
+
+**Abstract**
+
+IPLD Gateway is an extension of [PATH_GATEWAY.md](./PATH_GATEWAY.md) that enables lower level interaction with IPLD data structures under a specific path.
+
+This document describes the delta between [PATH_GATEWAY.md](./PATH_GATEWAY.md) and this gateway type.
+
+Summary:
+
+- Adds a new `/ipld/{cid}[/{segments}][?{params}]` subpath to the gateway
+- Defines a specification for parsing out extra parameters for individual path segments.
+- Describes how to map to and from `ipld://` URLs
+
+## HTTP API
+
+### `GET /ipld/{cid}[/{segments}][?{params}]`
+
+Resolve IPLD paths to some data.
+
+The path segments will be traversed with any parameters used to transform data along the way.
+
+The `format` query string parameter, or the `Accept` request header can be used to control the format which will be used to return the data.
+
+By default, data will be returned as `dag-json`
+
+<!--
+TODO: Cache control semantics?
+TODO: Add more details on how the traversal works?
+-->
+
+### `HEAD /ipld/{cid}[/{segments}][?{params}]`
+
+Resolves IPLD paths, and yields the same status code and headers as `GET`.
+
+### `POST /ipld/` and `POST /ipld/localhost/`
+
+Upload raw data to IPLD.
+The `body` of the request shall be parsed according to the `Content-Type` as IPLD data via standard encodings.
+`/localhost/` is used to support `POST ipld://localhost/` for uploading IPLD data to local nodes in web browsers that support it.
+
+The response will contain an `ipld://{cid}/` URL pointing at your data.
+
+<!--
+TODO: Only allow `/localhost/`? Get rid of `/localhost` from the spec if light clients with protocol handlers don't matter/
+-->
+
+### `PATCH /ipld/{cid}[/{segments}][?{params}]`
+
+This endpoint enables you to apply a [patch set](https://ipld.io/specs/patch/) to existing IPLD data.
+
+<!--
+TODO: Talk about content encoding
+TODO: Mention interaction with ADLs/Schemeas/Selectors
+-->
+
+The response will be an `ipld://` URL with your updated data.
+
+Note that the CID in the response URL will contain the same `segments` as in the request URL.
+e.g. if you patch data at `ipld://{cid1}/some/path/`, you will get back a URL that looks like `ipld://{updated cid}/some/path/`
+This enables you to make complex changes to a subtree in a dataset and get back a new root CID to use in your application.
+
+## HTTP Request
+
+### Request Headers
+
+#### `Accept` (request header)
+
+For `/ipld/{cid}/*` paths, the `Accept` header is used to indicate the encoding that should be used to return the data.
+This means that data initially encoded as `dag-json` will be transcoded to `dag-cbor` if the `application/vnd.ipld.dag-cbor` Accept header is used.
+
+- `application/json`: Interpret in the same way as `application/vnd.ipld.dag-json`.
```

Comment:
What if data is a valid JSON (and not DAG-JSON) added to ipfs with `json` codec (and not `dag-json`)?
Parsing it as dag-json will error, even tho it is a valid JSON.

@hacdias  and I discussed this edge case and ended up with requirement to check codec from CID, and if it is `json`, use generic JSON codec instead of `dag-json`.

Reply:
Interesting. Do you have text written up somewhere that I can copy paste here?

Reply:
@RangerMauve we have some wording here, but it may not be definitive:

https://github.com/ipfs/specs/blob/67fab21f17def74cfc1a596024d21e823df7acfa/http-gateways/PATH_GATEWAY.md?plain=1#L184-L187